### PR TITLE
cast_variant skips unsupported target type validation for empty inputs

### PR DESCRIPTION
**Pull request by abigalekim**
## Description
Solves the issue mentioned in https://github.com/rapidsai/cudf/issues/23361. cast_variant now rejects unsupported desired_type values consistently, regardless of whether the input column is empty or non-empty. Supported empty casts, such as empty variant values to INT32 or STRING, continue to return empty columns. This PR also refactors the UnsupportedTypesThrows test, which now covers both the empty-input (early-return) path and the non-empty-input (dispatch) path.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/io/experimental/variant_extract_test.cpp [hunk 1]
```diff
@@ -883,3 +883,35 @@ TEST_F(CastVariantTest, EmptyInput)
...
+{
+  // Unsupported target types must throw regardless of whether the input is empty or non-empty.
+  auto const stream = cudf::test::get_default_stream();
+
+  // Empty input: the early-return path must still validate the type.
+  auto const empty_values =
+    cudf::empty_like(cudf::structs_column_view{make_xyz_three_row_variant()}.child(1));
+  for (auto const id : {cudf::type_id::FLOAT32,
+                        cudf::type_id::FLOAT64,
+                        cudf::type_id::BOOL8,
```
Comment (vuule) [inline cpp/tests/io/experimental/variant_extract_test.cpp]: Don't we support floats now and are about to support bool? IMO we need some types here that we don't expect to ever support.

Comment (abigalekim) [reply cpp/tests/io/experimental/variant_extract_test.cpp]: Nevermind, I will edit this test to reflect unsupported types in the spec.

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/c1e0e9d8fb04533843b14b1888a0b778b05f988d

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/b0d37be3b2324f89cbf71e4bd61d1e4df4810dd4

File: cpp/src/io/parquet/experimental/variant_extract.cu [hunk 2]
```diff
@@ -886,6 +900,7 @@ std::unique_ptr<column> cast_variant(column_view const& values,
                                      rmm::device_async_resource_ref mr)
 {
   validate_variant_child(values);
+  cudf::type_dispatcher(desired_type, validate_variant_type_fn{});
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/variant_extract.cu]: Is there a way we can get rid of the type dispatcher here. It's generally quite expensive and the functor being dispatch seems trivial enough.

File: cpp/tests/io/experimental/variant_extract_test.cpp [hunk 3]
```diff
@@ -1032,11 +1033,9 @@ TEST_F(CastVariantTest, EmptyInput)
...
-TEST_F(CastVariantTest, CastToUnsupportedTargetThrows)
+TEST_F(CastVariantTest, UnsupportedTypeThrows)
 {
-  // cast_variant only supports INT8/16/32/64 and STRING targets. Every other target is rejected at
-  // compile-time dispatch on the requested output type, independent of the input bytes, so a single
-  // well-formed placeholder row triggers the same throw for all of them.
+  // Unsupported target types must throw regardless of whether the input is empty or non-empty.
   auto stream = cudf::test::get_default_stream();
   std::vector<uint8_t> const val{make_variant_primitive(variant_primitive_type::NULLVAL)};
   cudf::test::lists_column_wrapper<uint8_t> values(val.begin(), val.end());
```
Comment (vuule) [inline cpp/tests/io/experimental/variant_extract_test.cpp]: these are now unused; remove.

File: cpp/tests/io/experimental/variant_extract_test.cpp [hunk 4]
```diff
@@ -1053,11 +1052,24 @@ TEST_F(CastVariantTest, CastToUnsupportedTargetThrows)
                                        cudf::type_id::DECIMAL64,
                                        cudf::type_id::DECIMAL128};
 
+  // Empty input: the early-return path must still validate the type.
+  auto const empty_values =
+    cudf::empty_like(cudf::structs_column_view{make_xyz_three_row_variant()}.child(1));
+  for (auto const& id : ids) {
```
Comment (vuule) [inline cpp/tests/io/experimental/variant_extract_test.cpp]:
```suggestion
  for (auto const id : ids) {
```

**Review by vuule — COMMENTED** [review]
few small comments

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/df6e5619a28d8d8d5a5ae30e14182e298fb2a952

File: cpp/src/io/parquet/experimental/variant_extract.cu (hunk 2 above)
Comment (abigalekim) [reply cpp/src/io/parquet/experimental/variant_extract.cu]: yes, I have refactored it

Comment (vuule) [reply cpp/src/io/parquet/experimental/variant_extract.cu]: I think the comment only referred to `validate_variant_type_fn`, not the entire cast. I'm not sure that we want all casting in a single function.

**Comment by mhaseeb123** [comment]
Hi @abigalekim, please update the PR description to say what exactly we are doing here instead of just linking the issue 🙂. Please don't include the problem statement in the description, just what the PR exactly does and what does it fix in maybe a small 3-4 line paragraph

File: cpp/src/io/parquet/experimental/variant_extract.cu [hunk 5]
```diff
@@ -898,14 +855,44 @@ std::unique_ptr<column> cast_variant(column_view const& values,
...
+      d_null_mask);
+    CUDF_CUDA_TRY(cudaGetLastError());
+    auto const null_count =
+      num_rows - cudf::detail::count_set_bits(d_null_mask, 0, num_rows, stream);
+    return std::make_unique<column>(desired_type,
+                                    num_rows,
+                                    std::move(data),
+                                    null_count > 0 ? std::move(null_mask) : rmm::device_buffer{},
+                                    null_count);
+  };
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/variant_extract.cu]: nit: Consider moving to a standalone function instead of a lambda if possible

File: cpp/src/io/parquet/experimental/variant_extract.cu [hunk 6]
```diff
@@ -898,14 +855,44 @@ std::unique_ptr<column> cast_variant(column_view const& values,
...
 
-  return cudf::type_dispatcher(desired_type,
-                               cast_variant_fn{val_lists_device_view,
-                                               num_rows,
-                                               desired_type,
-                                               d_null_mask,
-                                               std::move(null_mask),
-                                               stream,
-                                               mr});
+  auto const do_primitive = [&]<typename T>() -> std::unique_ptr<column> {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/variant_extract.cu]: Perhaps worth adding `fixed_width` somewhere in the name since strings are not being handled here?

```suggestion
  auto const do_fixed_width_primitive = [&]<typename T>() -> std::unique_ptr<column> {
```

File: cpp/src/io/parquet/experimental/variant_extract.cu [hunk 7]
```diff
@@ -898,14 +855,44 @@ std::unique_ptr<column> cast_variant(column_view const& values,
...
+        cudf::strings::detail::make_strings_children(fn, num_rows, stream, mr);
+      auto const null_count =
+        num_rows - cudf::detail::count_set_bits(d_null_mask, 0, num_rows, stream);
+      return make_strings_column(num_rows,
+                                 std::move(offsets_column),
+                                 chars.release(),
+                                 null_count,
+                                 null_count > 0 ? std::move(null_mask) : rmm::device_buffer{});
+    }
+    default: CUDF_UNREACHABLE("validated above");
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/variant_extract.cu]: nit: Perhaps a less vague message here?

File: cpp/tests/io/experimental/variant_extract_test.cpp [hunk 8]
```diff
@@ -1053,11 +1050,24 @@ TEST_F(CastVariantTest, CastToUnsupportedTargetThrows)
...
+  auto const empty_values =
+    cudf::empty_like(cudf::structs_column_view{make_xyz_three_row_variant()}.child(1));
   for (auto const id : ids) {
-    SCOPED_TRACE(std::string{"target type_id: "} + std::to_string(static_cast<int32_t>(id)));
     EXPECT_THROW(static_cast<void>(cudf::io::parquet::experimental::cast_variant(
-                   values, cudf::data_type{id}, stream)),
-                 std::invalid_argument);
+                   *empty_values, cudf::data_type{id}, stream)),
+                 std::invalid_argument)
+      << "expected throw for type_id " << static_cast<int>(id) << " on empty input";
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/variant_extract_test.cpp]: nit: Please use `std::format` instead?

File: cpp/tests/io/experimental/variant_extract_test.cpp [hunk 9]
```diff
@@ -1053,11 +1050,24 @@ TEST_F(CastVariantTest, CastToUnsupportedTargetThrows)
...
+  }
+
+  // Non-empty input: the dispatch path must also throw for unsupported types.
+  auto col         = make_apache_variant(avf::primitive_int32);
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/variant_extract_test.cpp:1145]: 🚁🚁🚁 (iykyk 😄)

File: cpp/tests/io/experimental/variant_extract_test.cpp [hunk 10]
```diff
@@ -1053,11 +1050,24 @@ TEST_F(CastVariantTest, CastToUnsupportedTargetThrows)
...
+  auto const value = cudf::structs_column_view{col}.get_sliced_child(1, stream);
+  for (auto const id : ids) {
+    EXPECT_THROW(static_cast<void>(cudf::io::parquet::experimental::cast_variant(
+                   value, cudf::data_type{id}, stream)),
+                 std::invalid_argument)
+      << "expected throw for type_id " << static_cast<int>(id) << " on non-empty input";
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/variant_extract_test.cpp]: Same, please use `std::format`

File: cpp/src/io/parquet/experimental/variant_extract.cu (hunk 5 above)
Comment (abigalekim) [reply cpp/src/io/parquet/experimental/variant_extract.cu]: I think I have refactored this code out.

File: cpp/src/io/parquet/experimental/variant_extract.cu (hunk 2 above)
Comment (abigalekim) [reply cpp/src/io/parquet/experimental/variant_extract.cu]: I have re-refactored it to address this comment!

File: cpp/src/io/parquet/experimental/variant_extract.cu (hunk 7 above)
Comment (abigalekim) [reply cpp/src/io/parquet/experimental/variant_extract.cu]: have refactored this lambda out.

**Comment by abigalekim** [comment]
I have added the PR description as well!

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/704c9bda1e0f6c34d920550580b469f3c7e74f17

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/480db43f3f9d3cbd8cfa00ae8fccac1529e3a7e3

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/b35620a6c2360cbb0bbb528718d3f332453cfcd0

**Comment by abigalekim** [comment]
/ok to test https://github.com/rapidsai/cudf/pull/23462/commits/047d5bb61267e893f4c7cf65036e4e83b6e3d7f2

**Review by mhaseeb123 — APPROVED** [review]

**Review by vuule — APPROVED** [review]
